modules/internal/Atomics.chpl

```diff
@@ -249,16 +249,55 @@ module Atomics {
       return ret;
     }
 
+    /*
+       Stores `desired` as the new value, if and only if the original value is
+       equal to `expected`. Returns `true` if `desired` was stored, otherwise
+       updates `expected` to the original value.
+     */
+    inline proc compareExchange(ref expected:bool, desired:bool, param order: memoryOrder = memoryOrder.seqCst): bool {
+      extern externFunc("compare_exchange_strong", bool)
+        proc atomic_compare_exchange_strong(ref obj:externT(bool), ref expected:bool, desired:bool, succ:memory_order, fail:memory_order): bool;
+
+      var ret:bool;
+      on this {
+        var localizedExpected = expected;
+        ret = atomic_compare_exchange_strong(_v, localizedExpected, desired, c_memory_order(order), c_memory_order(readableOrder(order)));
+        expected = localizedExpected;
```

Comment:
Based on my reading of the Sutter stuff, I'm thinking this should be
```
if !ret then expected = localizedExpected;
```
i.e., only if the exchange is done should our callers should expect their `expected` to be updated from what they passed in.

Reply:
Based on outside conversation I believe this one is okay now.

Reply:
I think our conversation was basically that we'd be putting the same value back, so it should be fine, but as David pointed out, that can introduce comm for the module code, so it's worth fixing.

Reply:
(and I did change it in the latest commit)